Formalize 'evaluable parameter' (#2754)
modelica/ModelicaSpecification: chapters/classes.tex

\begin{example}
Direct and indirect cyclic dependency:
\begin{lstlisting}[language=modelica]
/* All of the following are illegal: */
parameter Real r = 2 * sin(r); // Depends directly on r.
parameter Real p = 2 * q;      // Indirect dependency on p via q = sin(p).
parameter Real q = sin(p);     // Indirect dependency on q via p = 2 * q.
\end{lstlisting}
\end{example}

\begin{example}
While declaration equations must not be cyclical, the use of initial equations can still introduce valid cyclic dependencies between parameters:
\begin{lstlisting}[language=modelica]
  parameter Real p = 2 * q; // This is the only declaration equation.
  parameter Real q(fixed = false);
initial equation
  q = sin(p); // OK, not a declaration equation.
\end{lstlisting}
\end{example}

\begin{example}
Breaking cyclic dependency.
\begin{lstlisting}[language=modelica]
model ABCD
  parameter Real A[n, n];
  parameter Integer n = size(A, 1);
end ABCD;

final ABCD a;
// Illegal cyclic dependency between size(a.A, 1) and a.n.

> ABCD b(redeclare Real A[2, 2] = [1, 2; 3, 4]);

Reviewer: I know the example is from the current specification, but I just realized that this example is actually illegal because A is not replaceable.

Reviewer 2: That's actually allowed according to 7.3.3 as far as I know, replaceable is only necessary when changing the type of the element (and changing array dimensions apparently does not count as changing the type).

Reviewer 3: Aren't dimensions part of the type in that sense? See https://specification.modelica.org/maint/3.4/Ch10.html#array-declarations

> The number of dimensions and the dimensions sizes are part of the type, and shall be checked for example at redeclarations

Reviewer: There is non-normative text in 7.3.3 saying:
>[*Redeclaration with the same type can be used to restrict variability and/or change array dimensions.*]

And according to 7.3, if no variability prefix is given the variability does not change either. I thought it would become a variable.

  function.
\end{itemize}

\subsection{Acyclic Bindings of Constants and Structural Parameters}\label{acyclic-bindings-of-constants-and-parameters}

The unexpanded binding equations for constants in the translated model must be acyclic after flattening.
Thus it is not possible to introduce equations for constants by cyclic dependencies.

> Structural parameters, on the other hand, cannot have cyclic dependencies by definition (see \cref{component-variability-prefixes-discrete-parameter-constant}).

Reviewer 4: The change in this section is not good at all. The current rule means that you can just look at parameter-bindings and see if there are any cycles or not - which is easy to understand, implement and explain. Restricting it to structural parameters makes it more complicated - and in particular I cannot see how it handles the example below if p and q were parameters.

Author: It's handled like this:
```
parameter Real p = 2 * q; /* Normal parameter, due to cycle with q. */
parameter Real q = sin(p); /* Normal parameter, due to cycle with p. */
```

The point is that we don't need acyclic binding for normal parameters, as their value isn't determined until initialization time.

I was going to add a non-normative text explaining that it's a good idea to associate a reason with every parameter being normal (as indicated by the comments in the listing above).  Then, consider a user making a variability error:
```
parameter Integer n = integer(p); /* Normal parameter, declaration equation has parameter variability. */
Real[n] x; /* Error: dimension is not structural expression. */
```
Now, the error message for the non-structural expression can include the reasons why first `n`, and then `p` are normal parameters.

Author: I have now added a non-normative text on generation of good error messages.

Author: Right, I should have made a comment when pushing [the fix](https://github.com/modelica/ModelicaSpecification/pull/2754/commits/8ba996136b0f3742e152b1faec57e3dc2f24739b).

@eshmoylova, would you agree to marking this as resolved?

Reviewer: Yes. Resolving.

\begin{lstlisting}[language=modelica]
import length = Modelica.Utilities.Strings.length; /* Pure external function */
parameter Integer n = length("Hello");             /* Structural parameter */
parameter Integer p = length("Hello")
  annotation(Evaluate = false);                    /* Normal parameter */
parameter Boolean b = false;

/* Variability analysis might cause evaluation of n: */
parameter Real x = if b and n > 3 then x else 0;

/* Variability analysis cannot cause evaluation of p: */
parameter Real y = if b and p > 3 then y else 0;
\end{lstlisting}
\end{example}

\begin{example}
Normal and structural parameters.
\begin{lstlisting}[language=modelica]
model ABCD
  parameter Real A[n, n]; // Requires that n is a structural parameter.
  parameter Integer n = size(A, 1);
end ABCD;

final ABCD a;
// Illegal since cyclic dependencies between size(a.A, 1) and a.n makes n a normal parameter.

ABCD b(redeclare Real A[2, 2] = [1, 2; 3, 4]);
// Legal since size of A is no longer dependent on n, making n a structural parameter.

ABCD c(n = 2); // Legal since n is no longer dependent on the size of A.

> parameter Real r = 2 * sin(r); // Normal parameter, since r = 2 * sin(r) is cyclic.

Reviewer 4: As before, I don't see why we want to allow these cases.

Author: Because it avoids the complicated corner cases of the old acyclic binding rule if we exclude normal parameters (including the non-fixed ones) from the analysis.  The importance of this sort of rule should be to make sure constants and structural parameters can be evaluated during translation – systems of equations in the initialization problem involving normal parameters is something we should consider normal in Modelica.

Reviewer 4: Systems of equations involving parameters may be normal, but writing them in terms of parameter bindings is error-prone and if pushed too far will lead to convoluted binding equations. If I see:
`  parameter Real r=2*sin(r); `
I would not be sure what was intended. If I see:
```
  parameter Real r(fixed=false);
initial equation
  r/2=sin(r);
```
I'm sure that it is intended to solve r from an initial equation. Basically one major reason we introduce initial equations (compared to the previous work-arounds) was to increase readability.

Author: I agree that the use of `initial equation` is nice in this case.  However, we're coming from the situation described in #2742 where binding equations could still end up in systems of equations due to interactions with non-fixed parameters (that don't use binding equations, and thereby circumvents the current acyclic binding rule).

I'd be fine with suggesting a warning to be issued for the case of a cycle of normal parameters where all have declaration equations.

Reviewer 5: I agree with Hans. Allowing cyclic bindings for normal parameters seem unnecessary and out of scope for the purposes of this MCP.

Author: The warning could also be combined with – or replaced by – making it an error if there is a cycle involving only parameter declaration equations (no holes in the cycle corresponding to non-fixed parameters' initial equations).

Reviewer: A warning does not clarify to me how to interpret that expression, as Hans said. One may choose to interpret it as ``r = 2 * sin(r.start)``.

Reviewer: This has now been resolved by not allowing the cycles, I believe.

Author: Yes, the current form of the example is essentially the same as before:
```
/* All of the following are illegal: */
parameter Real r = 2 * sin(r); // Depends directly on r.
…
```
Resolving.

\begin{example}
\begin{lstlisting}[language=modelica]
constant Real p = 2 * q;
constant Real q = sin(p); // Illegal since p = 2 * q, q = sin(p) are cyclical
\end{lstlisting}
\end{example}

\subsection{Component Variability Prefixes discrete, parameter, constant}\label{component-variability-prefixes-discrete-parameter-constant}

The prefixes \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant! of a component declaration
are called variability prefixes and define in which situation the
variable values of a component are initialized (see \cref{events-and-synchronization} and
\cref{initialization-initial-equation-and-initial-algorithm}) and when they are changed in transient analysis (= solution
of initial value problem of the hybrid DAE):
\begin{itemize}
\item
  A variable \lstinline!vc! declared with \lstinline!constant! prefix remains constant during transient analysis, with a value that is unaffected by the initialization problem (i.e., determined during translation).
\item
  % TODO: Add "structural parameter" to index.

>   A variable \lstinline!vc! is called a \firstuse{structural parameter} if all of the following applies:

Author: Good question.  I guess you are right about _n_, since _c_ was already taken for _constant_.

I fixed it so they at least have different names.  Personally, I don't really see the point of using different names at all.  Calling them all `v` or `x` allows the reader to concentrate on the definitions instead of thinking about what information the variable names are trying to convey.

>   A variable \lstinline!vc! is called a \firstuse{structural parameter} if all of the following applies:

Author: This conversation is very old and outdated in view of the last design meeting discussions.  Resolving.

  instants (see \cref{events-and-synchronization}).
\item
  A \emph{continuous-time} variable \lstinline!vn! may have a non-vanishing time
  derivative (\lstinline!der(vn)<>0! possible) and may also
  change its value discontinuously at any time during transient analysis
  (see \cref{events-and-synchronization}). If there are any discontinuities the variable is
  not differentiable.
\end{itemize}

\begin{nonnormative}
With every normal parameter, there is at least one reason why it isn't a structural parameter.
This information is useful to maintain in tools, as it allows generation of informative error messages when a violation of structural expression variability is detected.
For example:
\begin{lstlisting}[language=modelica]
parameter Integer n =
  if b then 1 else 2;        /* Normal parameter due to cycle with 'b'. */
parameter Boolean b = n > 3; /* Normal parameter due to cycle with 'n'. */
Real[n] x;                   /* Variability error: 'n' must be structural. */
\end{lstlisting}
Here, a good error message for the variability error can include the information that the reason for \lstinline!n! being a normal parameter is that it is in a cycle with \lstinline!b!.

> (Would structural and normal parameters have had different variability prefixes in their component declarations, it would have been possible to detect and report errors directly at the structural component declarations instead.)

Author: Actually not.  This part of the non-normative block wants help the user understand why parameters in Modelica can't be as easily described as the other component variabilities.

> (Would structural and normal parameters have had different variability prefixes in their component declarations, it would have been possible to detect and report errors directly at the structural component declarations instead.)

Reviewer: I don't like really this sentence (line 404 in classes.tex). Maybe it's the use of "would" that makes it sound to me as if we wish it was in the specification but alas. It just does't sit right me. It's like we are putting into the specification what we wish the specification was instead of just making it so. That's why I thought it was a note to self that accidentally got into the change. Thinking of how to rephrase this sentence, it also occurs to me that we don't want to have such prefixes. Whether a parameter is structural or normal might change depending on a modification in a component or while extending. Having a fixed prefix would prevent that.

> (Would structural and normal parameters have had different variability prefixes in their component declarations, it would have been possible to detect and report errors directly at the structural component declarations instead.)

Author: The purpose of the comment is to help the reader understand why it needs to be as complicated as the example
illustrates.  Surely, there are pros and cons of having the same prefix for two different things, but I can try to formulate it in a less wishful way.  I'm just too tired to think of something right now.  Maybe tomorrow.

> (Would structural and normal parameters have had different variability prefixes in their component declarations, it would have been possible to detect and report errors directly at the structural component declarations instead.)

Author: (I am not in a position to prepare commits due to Easter holidays, so I'll just present an idea for now.)

How about something like this?
> The need to explain why a parameter is normal is the price we pay for not needing to make an explicit choice between structural and normal parameter using different variability prefixes in the component declarations.

> (Would structural and normal parameters have had different variability prefixes in their component declarations, it would have been possible to detect and report errors directly at the structural component declarations instead.)

Author: This conversation is very old and outdated in view of the last design meeting discussions.  Resolving.

To see this, note that \lstinline!Evaluate = false! makes it a normal parameter by definition, and that \lstinline!Evaluate = true! would trigger the warning if the parameter is normal.
\end{nonnormative}

By the acyclic binding rule in \cref{acyclic-bindings-of-constants-and-parameters}, it follows that the value of a constant or structural parameter to be used in simplifications is possible to obtain by evaluation of a structural expression where values are available for all component subexpressions.

\begin{example}
A particularly demanding aspect of this evaluation is the potential presence of external functions.
Hence, if it is known that a parameter won't be used by a structural expression, a user can make it clear that the external function is not meant to be evaluated during translation by using \lstinline!Evaluate = false!:
\begin{lstlisting}[language=modelica]
import length = Modelica.Utilities.Strings.length; /* Pure external function */
parameter Integer n = length("Hello");             /* Structural parameter */
parameter Integer p = length("Hello")
  annotation(Evaluate = false);                    /* Normal parameter */
parameter Boolean b = false;

> /* Variability analysis might cause evaluation of n: */

Reviewer: What is *variability analysis*?

Author: The process of determining the variability of components and expressions.  This is needed to detect variability violations, and is now proposed to also be part of the classification of parameters as of structural or normal.

Reviewer: It is a good name. Is it defined already?

Author: No, and now it isn't present in the example anymore.  Resolving.

> parameter Real x = if b and n > 3 then x else 0;

Reviewer: It is not clear to me what the value of ``x`` is supposed to be. If ``b and n > 3 = true``, the equation would be ``x = x``. Is the tool supposed to solve it as such, then ``x`` in undetermined. Or are you supposed to convert it to ``x = x.start``?

Author: Sure the example should be improved to avoid this kind of questions.  The point of the example was just to discuss variability analysis, not how to actually solve the equations.

Fixed in a79c61b.

Author: Appears to have been fixed long time ago.  Resolving.

/* Variability analysis cannot cause evaluation of p: */
parameter Real y = if b and p > 3 then y else 0;
\end{lstlisting}
\end{example}

\begin{example}
Normal and structural parameters.
\begin{lstlisting}[language=modelica]
model ABCD

>   parameter Real A[n, n]; // Requires that n is a structural parameter.

Reviewer: As it currently stands, there is nothing that says n is required to be a structural parameter. I guess, that part would come later as stated in the PR description:
> Once we agree what these should be, it remains to update the specification to say structural expression rather than parameter expression everywhere something needs to be known at translation time.

Author: Yes.  If all those changes (a quick search gives 60 matches for _parameter expression_) were made now, I'm afraid it would distract attention from the key concepts.

Author: It was decided at the design meeting to not make these changes as part of this PR.  Resolving.

The prefixes \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant! of a component declaration
are called variability prefixes and define in which situation the
variable values of a component are initialized (see \cref{events-and-synchronization} and
\cref{initialization-initial-equation-and-initial-algorithm}) and when they are changed in transient analysis (= solution
of initial value problem of the hybrid DAE):
\begin{itemize}
\item
  A variable \lstinline!vc! declared with \lstinline!constant! prefix remains constant during transient analysis, with a value that is unaffected by the initialization problem (i.e., determined during translation).
\item
  % TODO: Add "structural parameter" to index.
  A variable \lstinline!vc! is called a \firstuse{structural parameter} if all of the following applies:
  \begin{itemize}
  \item
    It is declared with the \lstinline!parameter! prefix.
  \item
    It has \lstinline!fixed = true!.
  \item
    It does not have annotation \lstinline!Evaluate = false!.
  \item
    The expression of its declaration equation does not depend on \lstinline!vc! itself, neither directly nor via cyclic dependency on other parameters.

>     To satisfy this condition, dependencies shall be removed as needed by applying simplifications based on values of constants and other structural parameters that don't depend on \lstinline!vc! with \lstinline!Evaluate = true!.
>     % What about breaking cycles by simplification based on values of other evaluated structural parameters that don't have Evaluate = true?

Author: The definition would be better – easier to reason about and to implement consistently across tools – done this way:
```suggestion
    To satisfy this condition, dependencies shall be removed as needed by applying simplifications based on values of constants (except with \lstinline!Evaluate = false!)  and all other structural parameters that don't depend on \lstinline!vc!.
    That the value of a structural parameter is used for these simplifications does not mean that it has to be determined during translation, but if \lstinline!vc! is found to be a structural parameter, then a Modelica tool will be able to break all cycles involving \lstinline!vc! by making some (possibly none or all) of the other structural parameters determined during translation.
```

Author: Since several concerns have been raised already, but none of them regarding this change, I'm applying it now.  Then, we can discuss it further as part of the (new) current state of the proposal.

\subsection{Acyclic Bindings of Constants and Parameters}\label{acyclic-bindings-of-constants-and-parameters}

> For a constant or parameter \lstinline!v! with declaration equation, the unexpanded expression of the declaration equation in the flattended model must not depend on \lstinline!v! itself, neither directly nor indirectly via other variables' declaration equations.

Author: The term _unexpanded_ used here is something I'm not sure I understand what it means.  Is this parenthesis from the _Modelica DAE Representation_ appendix where it gets defined?
>   Expanding all class definitions (flattening the inheritance tree) and adding the equations and assignment statements of the expanded classes for every instance of the model.

If so, we should probably make that definition more prominent.  If not, can this rule be expressed using other (established) terms?  (@HansOlsson, I'm guessing that you are the most qualified person to answer this.)

Reviewer: I think expand usually means expanding all records operator records, and arrays to a set of scalars of primitive types. (At least, this is how it appears in Section 4.7 Balanced Models. The HasCycles example has cycles when you look at the arrays. However, if you expand the equations into the equations for scalars there are no cycles left.

Author: Sounds like a plausible interpretation.  If no other interpretation pops up, I'll try to address this by either adding _expand_ to the index or adding a link link directly to where it is explained.

Reviewer 4: Yes, unexpanded here was to ensure that we don't have to make a special case for x and y in:
```
final constant Real A [3 , 3] = [0 , 0 , 0; 1 , 0 , 0; 2 , 3 , 0];
parameter Real y [3] = A * y + ones (3) ;
parameter Real x [ n ] = cat (1 , {3.4} , x [1:( n -1) ]) ;
```
I'm not aware if there are any realistic use-cases for this. The benefit is both that we specifying the details (should some functions be inlined?), and that it doesn't require tools to scalarize equations needlessly to check this.

Author: Clarified in [commit](https://github.com/modelica/ModelicaSpecification/pull/2754/commits/b1b0dd4fe982e975720b40afdfd6e53ad1807c2b).  I didn't find any useful way of linking to the meaning of _expanded_, so I wrote it with more words instead, similar to the other places. OK?

Author: I take that as absence of objections.  Resolving.

  function.
\end{itemize}

\subsection{Component Variability Prefixes}\label{component-variability-prefixes-discrete-parameter-constant}

> The prefixes \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant! of a component declaration are called \firstuse[variability!prefix]{variability prefixes} and are the basis for defining in which situation the variable values of a component are initialized (see \cref{events-and-synchronization} and \cref{initialization-initial-equation-and-initial-algorithm}) and when they are changed during \firstuse{transient analysis} (i.e., solution of initial value problem of the hybrid DAE).

Author: This is the first use of the term _transient analysis_, but introducing the definition here looks odd to me.  The question is whether we'd really want to move the introduction somewhere else, considering that the term is only used in this chapter?

One alternative to consider would be to define both _initialization problem_ and _transient analysis_ after the definition of _flattening_ in the introduction: https://specification.modelica.org/master/introduction1.html#some-definitions

Reviewer: I do not implement initialization or simulation, so to me the distinction between "solution of initial value problem of the hybrid DAE" and initialization is not clear. I always thought they were the same. The "transient analysis" term is only used here, but it is a crucial concept in how/when to determine parameters. So if it is not the same as initialization, I would like this difference explained better. Other term that I perhaps confuse is flattening vs translation. We have the official definition of flattening (the translation of ...) but there are many places where we call it the translation phase and not the flattening phase. Are they the same or different? Do we have an overview of all the steps or at least the suggestion of how a model gets from Modelica code to a simulation, where we can clearly see how all the steps like flattening, translation, clock partitioning, initialization, transient analysis, simulation, etc are related?

Author: Sounds like a good idea, but I think we have to start thinking seriously about the scope creep in this PR.  How about opening a separate issue/PR about this, and revert the introduction of `\firstuse` here to avoid merge conflicts?

Reviewer 4: > I do not implement initialization or simulation, so to me the distinction between "solution of initial value problem of the hybrid DAE" and initialization is not clear. I always thought they were the same. The "transient analysis" term is only used here, but it is a crucial concept in how/when to determine parameters. 

The "transient analysis" is the phase _after_ the initialization; where you have "transients" (i.e. time-varying behavior). That's a fairly normal definition for circuit simulations (and is often contrasted with static simulations). I don't see that there's a clear separation between "transient analysis" and "simulation" for us.

The problem here seems to be:
"and when they are changed in transient analysis (= solution of initial value problem of the hybrid DAE):"
That was added in Modelica 1.2, extended slightly in 1.3 and has been basically unchanged since then.

Here 'solution of initial value problem' isn't related to 'initialization' but to simulation from a given initial value (the initial value problem cp. boundary value problem).

So there is no actual error, but it is not helpful as it is introducing one term in terms of another that isn't defined (and 'initial value problem' is used in two other places  - as far as I can see only 50% of them are correct).

I will make a PR clearing up this.

Reviewer 4: The new PR is #3146 (and clarified previous comment).

Author: @eshmoylova: OK to put faith in #3146 and remove the introduced use of `\firstuse` from the present PR?

Reviewer: OK, resolving, given that it would be merged/updated after #3146.

Author: Good.  The introduction of `\firstuse` has been reverted.

\subsection{Component Variability Prefixes discrete, parameter, constant}\label{component-variability-prefixes-discrete-parameter-constant}

> The prefixes \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant! of a component declaration are called \firstuse[variability!prefix]{variability prefixes}\index{component variability}\index{declared variability}\index{variability!declared|see{declared variability}} and define in which situation the variable values of a component are initialized (see \cref{events-and-synchronization} and \cref{initialization-initial-equation-and-initial-algorithm}) and when they are changed during transient analysis (i.e., solution of initial value problem of the hybrid DAE):

Reviewer 4: It might be a rest from the time when we considered different analysis for models (transient and initial).

> The prefixes \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant! of a component declaration are called \firstuse[variability!prefix]{variability prefixes}\index{component variability}\index{declared variability}\index{variability!declared|see{declared variability}} and define in which situation the variable values of a component are initialized (see \cref{events-and-synchronization} and \cref{initialization-initial-equation-and-initial-algorithm}) and when they are changed during transient analysis (i.e., solution of initial value problem of the hybrid DAE):

Author: A quick fix would be to make this a `\firstuse` definition:
>  The prefixes \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant! of a component declaration are called \firstuse[variability!prefix]{variability prefixes}\index{component variability}\index{declared variability}\index{variability!declared|see{declared variability}} and define in which situation the variable values of a component are initialized (see \cref{events-and-synchronization} and \cref{initialization-initial-equation-and-initial-algorithm}) and when they are changed during transient analysis (i.e., solution of initial value problem of the hybrid DAE):

The problem is that the term is only used very locally in the document.  Better ideas?

> The prefixes \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant! of a component declaration are called \firstuse[variability!prefix]{variability prefixes}\index{component variability}\index{declared variability}\index{variability!declared|see{declared variability}} and define in which situation the variable values of a component are initialized (see \cref{events-and-synchronization} and \cref{initialization-initial-equation-and-initial-algorithm}) and when they are changed during transient analysis (i.e., solution of initial value problem of the hybrid DAE):

Author: Short of better ideas, I'll begin by applying the `\firstuse` fix.  If needed we can then continue the discussion by commenting on the new use of `\firstuse`.  I'll resolve this once the `\firstuse` fix has been committed.

\begin{itemize}
\item
  A variable \lstinline!vc! declared with \lstinline!constant!\indexinline{constant} prefix remains constant during transient analysis, with a value that is unaffected by the initialization problem (i.e., determined during translation).
  This is called a \firstuse[---]{constant}, or \firstuse[constant!variable]{constant variable}\index{component variability!constant}.
\item

>   A variable \lstinline!ep! is called a \firstuse[parameter!evaluable]{evaluable parameter variable}\index{component variability!evaluable parameter} if all of the following applies:

Reviewer: Should there also be "or simply _evaluable parameter_" similar to the way it is for constants and normal parameters?

Author: Done.  OK?

Reviewer: OK, resolving.

of initial value problem of the hybrid DAE):
\begin{itemize}
\item
  A variable \lstinline!vc! declared with \lstinline!constant! prefix remains constant during transient analysis, with a value that is unaffected by the initialization problem (i.e., determined during translation).
\item
  % TODO: Add "structural parameter" and "evaluated parameter" to index.
  A variable \lstinline!vs! is called a \firstuse{structural parameter} if all of the following applies:
  \begin{itemize}
  \item
    It is declared with the \lstinline!parameter! prefix.
  \item
    It has \lstinline!fixed = true!.
  \item
    It does not have annotation \lstinline!Evaluate = false!.
  \item
    The declaration equation -- if any -- is given by a structural expression (\cref{structural-expressions}).

>     (Not having a declaration equation is similar to the expression being the default value for the component's type, which is constant.)

Reviewer: Could you explain this sentence? Is "the default value for the component's type" the same as the start value?

Author: OK, I might be mistaken here, so let's consider an example:
```
model M
  parameter Boolean p; // Fine: parameter (regardless of fixed) doesn't have to be modified.
  constant Boolean c; // Fine: constant doesn't have to be modified.
end M;
```
The comments express my current understanding of the role of declaration equations thought of as modifications.  The value of the unmodified Booleans is false, coming from the `start` of the pseudocode definition of `Boolean` in 4.8.3 _Boolean Type_.

(System Modeler gives a warning for `p`, but inconsistently silently accepts `c`.)

Author: In a different context, @qlambert-pro just pointed me to the section that rules out the constant without declaration equation.  Quite surprisingly, it's stated in 3.8.1 _Constant Expressions_, rather than when describing constant component declaration in 4.4.4 _Component Variability Prefixes discrete, parameter, constant_:
> Components declared as `constant` shall have an associated declaration equation with a constant expression, if the constant is directly in the simulation model, or used in the simulation model.

Reviewer: Here is another example:
```
type ThermodynamicTemperature = Real(start = 288.15);

model M
  parameter ThermodynamicTemperature T(start = 300);
...
end M;
```
The proposed text:
>(Not having a declaration equation is similar to the expression being the default value for the component's type, which is constant.)

suggests to me that the value of T should be 0. I am not sure if 288.15 can be considered  "the default value for the component's type" because it is modified. However, it would make more sense to take T.start (300) as the value of T, since this is the latest modification.

I am still not sure what "which is constant" refers to in the quoted sentence.

Author: I sincerely appreciate your effort to understand the proposal.  It really helps.

You are totally right regarding the role of `start` here for a `fixed = true` parameter without declaration equation.  It somehow slipped my mind that what we are talking about here is the special thing used, for example, in `Modelica.Electrical.Analog.Basic.Resistor` for the parameter `R`.  This isn't defined in either 3.8 (causing lack of symmetry with the case of `constant` without declaration equation) or 4.4.4 (which is the place I would try first if I didn't know where to find it), but in 8.6, item 3:
> Implicitly by using the `start` attribute for variables with `fixed = true`…
> - For a variable declared as `constant` or `parameter`, no equation is added to the initialization equations.

(The text looks outdated, since a constant is required to have a declaration equation.)

What it means here (see pushed change) is that `start` simply takes the role of the declaration equation in case it is missing.

It means that one can construct a normal parameter like this:
```
model M
  parameter Real T0 = 288.15 annotation(Evaluate = false); // Normal parameter, due to Evaluate = false.
  parameter Real T(start = T0); // Normal parameter, since no declaration equation and start isn't structural.
end M;
```

Author: I suppose this is considered first clarified, and then obsolete in view of the new fallback values.  Resolving.

      Medium.specificEnthalpy_pTX(p_start, T_start, X_start) else Medium.h_default
    annotation(Dialog(tab = "Initialization", enable = not use_T_start));
end PartialLumpedVolume;

> // Cycle for T_start and h_start, but still strucutral since cycle disappears

Reviewer 6: ```suggestion
// Cycle for T_start and h_start, but still structural since cycle disappears
```

\end{lstlisting}
\end{example}

\subsection{Conditional Component Declaration}\label{conditional-component-declaration}

A component declaration can have a \lstinline!condition-attribute!: \lstinline!if!~\emph{expression}.

\begin{example}
\begin{lstlisting}[language=modelica]
  parameter Integer level(min=1)=1;
  Motor motor;
  Level1 component1(J=J) if level==1 "Conditional component";
  Level2 component2 if level==2 "Conditional component";
  Level3 component3(J=component1.J) if level<2 "Conditional component";
  // Illegal modifier on component3 since component1.J is conditional
  // Even if we can see that component1 always exist if component3 exist
equation
  connect(component1$\ldots$, $\ldots$) "Connection to conditional component 1";
  connect(component2.n, motor.n) "Connection to conditional component 2";
  connect(component3.n, motor.n) "Connection to conditional component 3";
  component1.u=0; // Illegal
\end{lstlisting}
\end{example}

> The \emph{expression} must be a \lstinline!Boolean! scalar expression, and must be a parameter expression.

Reviewer 4: Shouldn't this be:
```suggestion
The \emph{expression} must be a \lstinline!Boolean! scalar expression, and must be an evaluable expression.
```
?

Author: No, but just because we decided to only apply the new terminology to array dimensions in this PR.  As soon as this PR is merged, we should set up a separate PR for this change.

Reviewer 4: Ok

Author: Good.  Resolving.

\begin{nonnormative}

> A parameter expression is required since it shall be evaluated at compile time.

Reviewer 4: ```suggestion
An evaluable expression is required since it shall be evaluated at compile time.
```

Author: Should to into same PR as the suggested change above.

Author: As it's a non-normative comment to the normative change that should go into a separate PR, it seems best to work out the two in conjunction.  When we get to do that, it's worth reconsidering the strong wording with _shall_ used inside of a non-normative section.  Maybe it's better to let the evaluable variability speak for itself, and not arbitrarily require that some evaluable expressions must be evaluated during translation…

We also have a similar strong formulation to look at for the `assert`-equation:
```
\begin{nonnormative}
A parameter expression is required for \lstinline!level! since it shall be evaluated at compile time.
\end{nonnormative}
```

Author: I take the approving review as a sign that we continue this discussion in a new PR.  Resolving.

A component declared with a condition-attribute can only be modified and/or used in connections.

\begin{nonnormative}
Adding the component and then removing it ensures that the component is valid.

If a connect equation defines the connection of a non-conditional component \lstinline!c1! with a conditional component \lstinline!c2! and \lstinline!c2! is de-activated, then \lstinline!c1! must still be a declared element.

There are annotations to handle the case where the connector should be connected when activated, see \cref{annotations-for-the-graphical-user-interface}.
\end{nonnormative}

\section{Component Variability}\label{component-variability}

As briefly mentioned in \cref{component-variability-prefixes-discrete-parameter-constant}, the component variability prefixes are the basis for defining \firstuse{component variability}\index{declared variability}\index{variability!declared|see{declared variability}}.
Combined with some other information about the components and analysis of expression variability (\cref{variability-of-expressions}), they define the component variabilities as follows:
\begin{itemize}
\item

>   A variable \lstinline!vc! declared with \lstinline!constant!\indexinline{constant} prefix does not change during simulation, with a value that is unaffected even by the initialization problem (i.e., determined during translation).

Reviewer 4: I would drop "determined during translation":
```suggestion
  A variable \lstinline!vc! declared with \lstinline!constant!\indexinline{constant} prefix does not change during simulation, with a value that is unaffected even by the initialization problem.
```
Since one could have it bound to an external function, that is evaluated as part of running the model (seen as before initialization at least in Dymola).

I realize that with the change of including constants as part of evaluable, it must be possible to evaluate it - but there's still no need to _actually evaluate it_ during translation.

Author: It says _determined_, not _evaluated_, so I think that the clarification is helpful without being misleading.

Author: I take the approving review as a sign of this being OK as is.  Resolving.

parameter Boolean b = false;

/* Variability analysis might cause evaluation of n: */

> parameter Real x = if b and n > 3 then 1 - x else 0;

Reviewer: With the latest revision would this example be allowed still?

Author: Again, GitHub seems to be doing strange things to your comments, so I'm quoting what I'm answering to:
> It is a good name. Is it defined already?

I think you are referring to _variability analysis_?  No, it's not defined as far as I know, it just comes natural to me.  Of course, we should't use undefined terms like this, so I suggest we explain it under 3.8 _Variability of Expressions_.

Author: Now the other comment:
> With the latest revision would this example be allowed still?

Good catch!  The concept driving parameter evaluation has changed from variability analysis to fulfillment of the acyclic binding rule.  Better now?

To see this, note that \lstinline!Evaluate = false! makes it a non-evaluable parameter by definition, and that \lstinline!Evaluate = true! would trigger the warning if the parameter is non-evaluable.
\end{nonnormative}

\begin{nonnormative}
With every non-evaluable parameter, there is at least one reason why it isn't an evaluable parameter.
This information is useful to maintain in tools, as it allows generation of informative error messages when a violation of evaluable expression variability is detected.
For example:
\begin{lstlisting}[language=modelica]
  parameter Integer n =
    if b then 1 else 2;    // Non-evaluable parameter due to variability of b.
  parameter Boolean b(fixed = false);
                           // Non-evaluable parameter due to fixed = false.
  Real[n] x;               // Variability error: n must be evaluable.
initial equation
  b = n > 3;
\end{lstlisting}
Here, a good error message for the variability error can include the information that the reason for \lstinline!n! being a non-evaluable parameter is that it has a dependency on the non-evaluable parameter \lstinline!b!.

> (Would evaluable and non-evaluable parameters have had different variability prefixes in their component declarations, it would have been possible to detect and report errors directly at the evaluable component declarations instead.)

Reviewer 4: To me this seems like a consideration for a different design that shouldn't be present in the specification.

Author: Yeah, I don't remember how it ended up here; maybe it was meant more as a clarification to reviewers than to end users.  I'll remove it.

Author: Removed.  Resolving.
